### ISSUE TITLE
Timeout support not backwards compatible with Ruby 1.9

### ISSUE THREAD

**Issue by james-masson** [hashicorp/vault-ruby#17]
Hi,

It looks like https://github.com/hashicorp/vault-ruby/commit/4eff7cbe02b26a3601e50f056671f4e60c9369c9  broke compatibility with Ruby 1.9

Why is this important?

It means that you can't use Hiera-Vault   (https://github.com/jsok/hiera-vault)  integration with PuppetServer (https://github.com/puppetlabs/puppet-server) - which uses jRuby 1.9
Many RHEL variants still use Ruby 1.9 (or even 1.8) - so it's not an edge case.

While I will ask Puppetlabs about newer Ruby/jRuby support in PuppetServer, I wouldn't expect anything soon as likely it will mean them dropping backward compatibility.

In the meantime, I'm building a custom vault-ruby gem for our use case. I'm not saying there's a nice way out of this, or that you should actively support Ruby 1.9 - just that this change might cause some collateral damage!

thanks

James M

**Comment by justincampbell** [comment]
@james-masson Can you point out where exactly in the code it's broken? I think we could make this gem work for both Ruby 2+ and JRuby 1.9-mode. I've opened a PR to run CI against JRuby in #18. Any help there would be appreciated!

**Comment by james-masson** [comment]
From your acceptance tests....

```
  1) Vault::Logical#delete deletes the secret
     Failure/Error: expect(subject.read("secret/delete")).to be(nil)
     NameError:
       uninitialized constant Net::ReadTimeout
     # ./lib/vault/client.rb:211:in `rescue in request'
     # ./lib/vault/client.rb:108:in `request'
     # ./lib/vault/client.rb:59:in `get'
     # ./lib/vault/api/logical.rb:27:in `read'
     # ./spec/integration/api/logical_spec.rb:39:in `block (3 levels) in <module:Vault>'

.....
```
and many more similar errors....

To reproduce:
* check out master
* ```rvm use 1.9``` - or equivalent
* ```bundle install```
* ```bundle exec rake```

Whereas with ```rvm use 2.0```
```
Finished in 0.499 seconds (files took 0.17571 seconds to load)
90 examples, 0 failures, 9 pending
```

As to what the backwards compatible solution is, I'm not sure - I'm just building from 
https://github.com/hashicorp/vault-ruby/commit/e3a42d298240affc0a83faec24cea3ad3e2e7125  + my TLS patch.

**Comment by sethvargo** [comment]
Hi @james-masson 

Thank you for opening an issue. Ruby 1.9 [reached EOL in February 2015](https://www.ruby-lang.org/en/news/2014/01/10/ruby-1-9-3-will-end-on-2015/). It has significant security vulnerabilities that will no longer be patched, and many tools are quickly [moving away from it](https://www.chef.io/blog/2014/11/25/ruby-1-9-3-eol-and-chef-12/). I understand that older distros still ship old versions of Ruby, but vault-ruby really requires Ruby 2.0+. I will add a note to the README to make this clearer.

There are a number of gaping security vulnerabilities in Ruby 1.9. Given that Vault is tool for security, I cannot justify supporting EOL versions of Ruby. It is possible, for example, in Ruby 1.9 to highjack the SSL connection by posing as a Vault server and then intercept the token from the connection. This is because Ruby automatically _downgrades_ it's SSL cipher suite to the lowest-common-denominator. 

All of that being said, I have fixed this issue and added 1.9 to the testing matrix. I'll do my best to maintain compatibility with Ruby 1.9, but please know that you assume significant security vulnerabilities in choosing to run Ruby 1.9 for which we are not responsible.
